grpc_stdio.go

```diff
@@ -0,0 +1,196 @@
+
+	for {
+		// Read our data
+		select {
+		case data.Data = <-s.stdoutCh:
+			data.Channel = plugin.StdioData_STDOUT
+
+		case data.Data = <-s.stderrCh:
+			data.Channel = plugin.StdioData_STDERR
+
+		case <-srv.Context().Done():
+			return nil
+		}
+
+		// Not sure if this is possible, but if we somehow got here and
+		// we didn't populate any data at all, then just continue.
+		if len(data.Data) == 0 {
+			continue
+		}
+
+		// Send our data to the client.
+		if err := srv.Send(&data); err != nil {
+			return err
+		}
+	}
+}
+
+// grpcStdioClient wraps the stdio service as a client to copy
+// the stdio data to output writers.
+type grpcStdioClient struct {
+	log         hclog.Logger
+	stdioClient plugin.GRPCStdio_StreamStdioClient
+}
+
+// newGRPCStdioClient creates a grpcStdioClient. This will perform the
+// initial connection to the stdio service. If the stdio service is unavailable
+// then this will be a no-op. This allows this to work without error for
+// plugins that don't support this.
+func newGRPCStdioClient(
+	ctx context.Context,
+	log hclog.Logger,
+	conn *grpc.ClientConn,
+) (*grpcStdioClient, error) {
+	client := plugin.NewGRPCStdioClient(conn)
+
+	// Connect immediately to the endpoint
+	stdioClient, err := client.StreamStdio(ctx, &empty.Empty{})
+
+	// If we get an Unavaiable error, this means that the plugin isn't
+	// updated and linking to the latest version of go-plugin that supports
+	// this. We fall back to the previous behavior of just not syncing anything.
+	if status.Code(err) == codes.Unavailable {
+		log.Warn("stdio service not available, stdout/stderr syncing unavailable")
+		stdioClient = nil
+		err = nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &grpcStdioClient{
+		log:         log,
+		stdioClient: stdioClient,
+	}, nil
+}
+
+// Run starts the loop that receives stdio data and writes it to the given
+// writers. This blocks and should be run in a goroutine.
+func (c *grpcStdioClient) Run(stdout, stderr io.Writer) {
+	// This will be nil if stdio is not supported by the plugin
+	if c.stdioClient == nil {
+		c.log.Warn("stdio service unavailable, run will do nothing")
+		return
+	}
+
+	for {
+		c.log.Trace("waiting for stdio data")
+		data, err := c.stdioClient.Recv()
+		if err != nil {
+			if err == io.EOF ||
+				status.Code(err) == codes.Unavailable ||
+				err == context.Canceled {
+				c.log.Warn("received EOF, stopping recv loop", "err", err)
+				return
+			}
+
+			c.log.Error("error receiving data", "err", err)
+			continue
+		}
+
+		// Determine our output writer based on channel
+		var w io.Writer
+		switch data.Channel {
+		case plugin.StdioData_STDOUT:
+			w = stdout
+
+		case plugin.StdioData_STDERR:
+			w = stderr
+
+		default:
+			c.log.Warn("unknown channel, dropping", "channel", data.Channel)
+			continue
+		}
+
+		// Write! In the event of an error we just continue.
+		c.log.Trace("received data", "channel", data.Channel.String(), "len", len(data.Data))
```

Comment:
These `waiting` and `received` log messages within the receive loop feel like they might be too chatty for busy clients.

Reply:
I changed this to a `c.log.IsTrace()` check prior. We don't normally ever default trace logs (even in dev I think we normally do Debug level), and I've been using them more since in a pinch of debugging they've been helpful.

I'm fine removing this but felt it could be useful in debugging why stdio might not be working if that ever happened cause you can see on both sides the data receival/send.